docs/src/index.md

```diff
@@ -27,7 +27,271 @@ this library has a few goals:
+p2.parse("t ")
+// res3: Either[Error, Tuple2[String, Char]] = Right((,t))
+
+/* OrElse */
+
+val p3: Parser[AnyVal] = alpha | sp
+
+p3.parse("t")
+// res4: Either[Error, Tuple2[String, AnyVal]] = Right((,t))
+p3.parse(" ")
+// res5: Either[Error, Tuple2[String, AnyVal]] = Right((,()))
+```
+
+## Repeating parsers
+
+Sometimes we need something to repeat zero or more types. The cats-parse have `rep` and `rep0` methods for repeating values. `rep` means that the parser must be successful *at least one time*. `rep0` means that the parser output might be empty.
+
+```scala
+import cats.data.NonEmptyList
+import cats.parse.Rfc5234.alpha
+import cats.parse.Parser
+
+val p1: Parser[NonEmptyList[Char]]  = alpha.rep
+val p2: Parser0[List[Char]] = alpha.rep0
+
+p1.parse("")
+// Left(Error(0,NonEmptyList(InRange(0,A,Z), InRange(0,a,z))))
+p2.parse("")
+// Right((,List()))
+p2.parse("something")
+// Right((,List(s, o, m, e, t, h, i, n, g)))
+```
+
+Notice the types of parsers. `Parser` type always means some non-empty output and the output of `Parser0` might be empty.
+
+## Parsers with empty output
+
+Some parsers never return a value. They have a type `Parser0`. One might get this type of parser when using `rep0` or `.?` methods.
+
+```scala
+import cats.parse.Rfc5234.{alpha, sp}
+import cats.parse.Parser
+
+val p: Parser[String] = (alpha.rep <* sp.?).rep.map { words => 
+  words.map(charsList => charsList.toList.mkString("")).toList.mkString(" ")
+}
+
+p.parse("hello world")
+// res0 = Right((,hello world))
+```
+
+Notice the type we got - `Parser[String]`. That is because we have `rep` outside and our `alpha.rep` parser with `Parser` type is on the left side of the clause. But what if we want to parse strings with spaces at the beginning?
+
+```scala
+val p = (sp.? *> alpha.rep <* sp.?).rep.map { words => 
+  words.map(charsList => charsList.toList.mkString("")).toList.mkString(" ")
+}
+```
+
+We will get an error `value rep is not a member of cats.parse.Parser0`. This happens since we have the left-side parser as optional in `sp.? *> alpha.rep <* sp.?` clause. This clause has a type `Parser0` which can't be repeated.
+
+But this parser can't be empty because of `alpha.rep` parser, and we know it. For these types of parsers we need to use `with1` wrapper method on the *left side* of the clause.
+
+```scala
+import cats.parse.Rfc5234.{alpha, sp}
+import cats.parse.Parser
+
+
+val p: Parser[String] = (sp.?.with1 *> alpha.rep <* sp.?).rep.map { words => 
+  words.map(charsList => charsList.toList.mkString("")).toList.mkString(" ")
+}
+
+p.parse("hello world")
+// res0: Either[Error, Tuple2[String, String]] = Right((,hello world))
+p.parse(" hello world")
+// res1: Either[Error, Tuple2[String, String]] = Right((,hello world))
+```
+
+If we have multiple `Parser0` parsers before the `Parser` - we'd need to use parenthesis like this:
+`(sp.? ~ sp.?).with1 *> alpha.rep`.
+
+## Error handling
+
```

Comment:
there are two kinds of errors:

1. an error that has consumed 0 characters (epsilon failure)
2. an error that has consumer 1 or more characters (halting failure).

I think it is important to mention that. `orElse` or `|` allows recovering from type 1, but not type 2 errors. `.backtrack` converts type 2 errors into type 1, but we want to generally minimize that for performance and locating errors.

Reply:
In the backtrack docs there is a phrase "This method converts arresting failures into epsilon failures". Is arresting = halting?

Reply:
yeah, sorry. I forgot the terminology used in the docs (I think @non came up with that but I always seem to wind up recalling halting).

https://github.com/typelevel/cats-parse/blob/714a9df6ac458ba14d654ef4d5c620452db02cb3/core/shared/src/main/scala/cats/parse/Parser.scala#L43

Erik wrote some really nice docs, so let's stick to his terminology.